### PR TITLE
YANG & NETCONF XML parser support empty strings `<tag></tag>`  and strings in self-closing tags `<tag />`

### PR DESCRIPTION
**Pull request by riccardo-negri**
Issue which we experienced: 
```
2026-05-20T15:05:30.316722Z DEBUG netgauze_yang_push::cache::actor: sent error response to requester due to fetch failure peer=10.190.64.79:10100 subscription_id=14 router_content_id="EMPTY" target={ datastore: EMPTY, datastore-xpath-filter: EMPTY } error=failed to connect to netconf server: Error in XML parsing the <rpc-reply> message: Recoverable
```

Now we do not experience it anymore:
```
2026-05-21T08:25:39.396008Z  INFO netgauze_yang_push::cache::fetcher: YANG Library fetched from device host=10.190.64.79:830 peer=10.190.64.79:10100 subscription_id=24 router_content_id="bcda49ca12973c1644ebbe8f20cf72d98b01c76c1425b0dafff3f27f4deef7f3" target={ datastore: ietf-datastores:operational, datastore-xpath-filter: /if:interfaces-state/if:interface/bbf-xponani:ani/bbf-xpon-ani-state:running-info } cached_content_id="bcda49ca12973c1644ebbe8f20cf72d98b01c76c1425b0dafff3f27f4deef7f3" schema_count=352
2026-05-21T08:25:39.396762Z  INFO netgauze_yang_push::cache::actor: fetched yang library from device successfully subscription_id=24 router_content_id="22934" target={ datastore: ietf-datastores:operational, datastore-xpath-filter: /if:interfaces-state/if:interface/bbf-xponani:ani/bbf-xpon-ani-state:running-info } modules_count=352 cached_content_id="bcda49ca12973c1644ebbe8f20cf72d98b01c76c1425b0dafff3f27f4deef7f3"
```

### CODE REVIEW

**Comment by ahassany** [comment]
I cannot accept this PR, a lot of changes that don't make sense. The original problem if I understood correctly, that `<purpose />` and `<purpose></purpose>` are not accepted, right? This behavior is tested in `test_tag_string_returns_empty_string_for_self_closing_tag` and `test_tag_string_returns_empty_string_for_explicit_empty_tag` which is correct way of verifying the bug and making sure the fix removes it. 

However, what is the point of the other tests? I ran them and they were passing before and after your PR? are they testing anything that was not tested before? Do they it reveal a bug? Are you just retesting quick-xml behavior?  or just AI generated slop?

And for the error change, what's the point of the new `MixedContent` error? the previous error explained exactly what the parser is expecting to get and what is got instead .. that's much clearer to the user I think.

Also, why the yang-push crate has changes in a PR about netconf?

Please, make your solution focused on actually solving a problem or adding a feature, and don't introduce noise that just takes plenty of time to review and just few minutes for AI to generate!

File: crates/netconf-proto/src/xml_utils.rs [hunk 1]
```diff
@@ -531,13 +530,13 @@ impl<'a, R: io::BufRead> XmlParser<'a, R> {
     }
 
     #[inline]
-    pub fn parent_has_child(&self) -> bool {
+    pub fn is_empty_element(&self) -> bool {
```
Comment (ahassany) [inline crates/netconf-proto/src/xml_utils.rs]: Why this function name changed? it doesn't make sense for this PR!

**Comment by riccardo-negri** [comment]
> I cannot accept this PR, a lot of changes that don't make sense. The original problem if I understood correctly, that `<purpose />` and `<purpose></purpose>` are not accepted, right? This behavior is tested in `test_tag_string_returns_empty_string_for_self_closing_tag` and `test_tag_string_returns_empty_string_for_explicit_empty_tag` which is correct way of verifying the bug and making sure the fix removes it.
> 
> However, what is the point of the other tests? I ran them and they were passing before and after your PR? are they testing anything that was not tested before? Do they it reveal a bug? Are you just retesting quick-xml behavior? or just AI generated slop?
> 
> And for the error change, what's the point of the new `MixedContent` error? the previous error explained exactly what the parser is expecting to get and what is got instead .. that's much clearer to the user I think.
> 
> Also, why the yang-push crate has changes in a PR about netconf?
> 
> Please, make your solution focused on actually solving a problem or adding a feature, and don't introduce noise that just takes plenty of time to review and just few minutes for AI to generate!

Hi Ahmed, thanks for the review. Yes, your understading of the bug is correct.
Let me address each point:

- The extra tests: these were added because the bug revealed that empty/self-closing element handling was under-tested as a whole. 
  While open(), close(), and maybe_open() themselves weren't modified, there were no existing tests covering the full lifecycle of opening,  interacting with, and closing empty elements: in fact my initial fix was leaving the parser in a bad state and this was not catched by the existing unit tests. 
  That said, I understand the concern: they mainly test unchanged code and pass before and after the fix. 
  I'm happy to remove them from this PR and submit them separately if you think they still add value as coverage, or drop them entirely.
  I will anyway remove test_parent_has_child_false_for_empty_tag() and test_tag_string_on_mixed_content_fails() because I noticed now that they are duplicates.

- The MixedContent error: I disagree here. 
  The fix requires splitting the End | Start | Empty match arm — End must now return Ok(accumulator) instead of erroring. 
  For the remaining Start | Empty case, MixedContent is more precise than the generic WrongToken: it uses correct XML terminology and it clearly communicates  why the input is invalid (child elements inside a leaf) rather than just what token was found. I'd like to keep this. Let me know if you have any concerns about it.

- The yang-push change: I understand how it is unrelated. However I kept it in a different commit on purpose and decided to include it in this PR because the debug message was misleading in finding and understanding the Netconf bug.

- The rename: You're right that is_empty_element() is semantically wrong: it returns true when the element is opened with Start (i.e., NOT empty), which is inverted. 
  However, the original parent_has_child() is also misleading: it doesn't check whether children exist, it checks whether the parent element's tag type allows children (i.e., was opened with Start rather than Empty). Both names misrepresent what the function does.
  I propose renaming it to element_expects_content() instead, which accurately describes the semantics: "was the current parent opened with a Start event, meaning the parser should expect content/children before a closing End event?"

I have removed the duplicated tests and renamed is_empty_element().

Thanks

**Comment by ahassany** [comment]
Please focus the PR on the actual bug and it's fix. If you want to introduce other changes, put them on separate PRs and we can discuss separately about these changes.  This couldn't been a 5min PR to review and merge, with the additional unnecessary changes it took couple hours from my time.

**Comment by riccardo-negri** [comment]
> Please focus the PR on the actual bug and it's fix. If you want to introduce other changes, put them on separate PRs and we can discuss separately about these changes. This couldn't been a 5min PR to review and merge, with the additional unnecessary changes it took couple hours from my time.

Dear Ahmed, I am very sorry to hear that those additional changes have made you take that much extra time. I would have not anticipated it due to them not adding functionalities or introducing big changes.

I reduced this MR just to the Netconf bug fix and the two related tests. I hope it will not take too much of your time to review it. 

Thank you!

**Comment by ahassany** [comment]
Thank you @riccardo-negri, if you feel the other parts need to be in NetGauze, feel free to push them as separate PRs
